reco_utils/aml/wide_deep.py

```diff
@@ -0,0 +1,136 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+"""
+AzureML Hyperdrive entry script for wide-deep model
+"""
+import argparse
+import os
+import shutil
+import sys
+
+import papermill as pm
+import tensorflow as tf
+print("TensorFlow version:", tf.VERSION)
+
+try:
+    from azureml.core import Run
+    run = Run.get_context()
+except ImportError:
+    run = None
+
+from reco_utils.common.constants import (
+    DEFAULT_USER_COL,
+    DEFAULT_ITEM_COL,
+    DEFAULT_RATING_COL
+)
+
+
+NOTEBOOK_NAME = os.path.join(
+    "notebooks",
+    "00_quick_start",
+    "wide_deep_movielens.ipynb"
+)
+OUTPUT_NOTEBOOK = "wide_deep.ipynb"
+SUPPORTED_METRICS = {
```

Comment:
I think this list of metrics can be difficult to maintain. When we add new metrics someone needs to remember to modify this file too. Do we need SUPPORTED_METRICS at all? 
I think it is reasonable to expect the user of the repo to pass the correct name of the metric.

Reply:
@anargyri That is fair point, but at the same time the current script helps to early-fail (raise an error before we train the model) which is useful especially when a user run remote-compute related work, like AzureML.

Would it be make sense to move this checking to the Notebook which actually imports those metrics and uses? There, the code maintenance would be easier as we will update the module import too.

Reply:
I think that would just move the problem to another file. But now you raise another issue, i.e. the import statement in this notebook needs to be maintained too. Also RANKING_METRICS, RATING_METRICS appear in the notebook. Now imagine maintaining this for every notebook that uses hyperdrive. So probably it is better to have an `import *` instead (even though that is discouraged by PEP8, but in this case it may be better) and remove RANKING_METRICS, RATING_METRICS.

That said, I agree that an early check that the metric exists is good to have inside the .py script, because it saves time. What could be done (instead of SUPPORTED_METRICS) is a try statement with a call to `callable(metric)` for each of the tracked metrics.

@miguelgfierro any thoughts?

Reply:
I guess your point applies to all notebooks. They all just use pre-selected metrics (currently, we use all metrics) and calculate each, like this:
```
from reco_utils.evaluation.python_evaluation import (rmse, mae, rsquared, exp_var, map_at_k, ndcg_at_k, precision_at_k, recall_at_k, get_top_k_items)
...
eval_map = map_at_k(test, all_predictions, col_prediction='prediction', k=TOP_K)
eval_ndcg = ndcg_at_k(test, all_predictions, col_prediction='prediction', k=TOP_K)
eval_precision = precision_at_k(test, all_predictions, col_prediction='prediction', k=TOP_K)
eval_recall = recall_at_k(test, all_predictions, col_prediction='prediction', k=TOP_K)

print("MAP:\t%f" % eval_map,
      "NDCG:\t%f" % eval_ndcg,
      "Precision@K:\t%f" % eval_precision,
      "Recall@K:\t%f" % eval_recall, sep='\n')
```
If we introduce new metrics, we anyway need to update those all.

But anyway I like to generalize if there is possible way to do that.
The thing is, for AzureML use-case, we have to somehow pass which metrics we want to track. We'll don't want to track all the metrics when we do 100 runs of Hyperdrive, especially when the data is big and thus ranking take so long.

Any thoughts? @anargyri @miguelgfierro

Reply:
That is a good point; we do the same elsewhere. Although I don't think that we maintain a list of SUPPORTED_METRICS anywhere; we refer to each metric explicitly by name. 

So I think what you said (moving the check to the notebook from the python script) is a better option. Another option is to just rely on the programmer to pass a valid METRICS list as an argument to the script.

Reply:
In the [benchmarks](https://github.com/Microsoft/Recommenders/blob/benchmark_feb/notebooks/03_evaluate/comparison_movielens.ipynb) (which is work in progress atm) I define which metrics go with which algo. I would say this is more explicit (which is one of our practices). So I would say that to move the check to the notebook makes sense

Reply:
Ok, in the next commit (which I'm going to push soon), the python script doesn't check them and expects users pass valid metrics instead. In the notebooks RANKING_METRICS and RATING_METRICS will contain metrics function names, which are expected to be the same as function.__name__ and thus the actual function will be called by locals()['name'] later cells. Will this better? @anargyri @miguelgfierro

The reason of using those lists is to give users option to select which metrics to track when use AzureML hyperdrive. Ranking takes more time, and the more metrics they use the longer the evaluation takes which end up more cost of cluster compute.

I tried to pass function directly from aml_notebook to model notebook, but papermill complains about "cannot serialize function".